src/reflect/scala/reflect/internal/tpe/TypeMaps.scala

```diff
@@ -236,10 +238,18 @@ private[internal] trait TypeMaps {
 
     /** Map this function over given list of symbols */
     def mapOver(origSyms: List[Symbol]): List[Symbol] = {
+      val firstChange = firstChangedSymbol(origSyms)
       // fast path in case nothing changes due to map
-      if (noChangeToSymbols(origSyms)) origSyms
-      // map is not the identity --> do cloning properly
-      else cloneSymbolsAndModify(origSyms, TypeMap.this)
+      if (firstChange < 0) origSyms
+      else {
+        // map is not the identity --> do cloning properly
+        val cloned = cloneSymbols(origSyms)
+        // but we don't need to run the map again on the unchanged symbols
+        cloned.splitAt(firstChange) match {
+          case (before, after) =>
+            before ::: mapList(after)(_ modifyInfo this)
```

Comment:
we use `applyToSymbolInfo` in `noChangeToSymbols`/`firstChangedSymbol`, but just `apply` here. Is that wrong? I didn't change it, but I wouldn't be surprised if it caused a bug. (I'll go looking.)

Reply:
You could avoid the split and ::: here and avoid the intermediate collection. - - if you had a mapList(skipN,...) where it just copies the first N across without change -

Reply:
Does applyToSymbolInfo and modilfyInfo - can we apply the clone and map in the same pass?

Reply:
I mean, the transformation I want is `cloneSymbolsAndModifySomeTail`. I'm not sure how hot this is and if it's worth it in the cost of code clarity. I'll see what I can profile.

You can't clone and modify in one pass, because the cloned symbols' infos need to have references to the original symbol rewritten. There may be some way to fuse this, but the amount of control that individual type maps have over their recursion makes that difficult.

Reply:
Yeah, I have no idea how hot this is - I struggled to get my head round this code.

Reply:
I mean, type maps get called into a _lot_, is my guess. adding stats tracking to these methods isn't easy because they're all recursive. I'll go figure out if I can get a profiler to count the non-recurrent calls. Although, the number of times through this method would be a good statistic as well. let's see.

I'll see if I can make a micro benchmark for trying some of the hotter type maps out. The bespoke half-list-mapping function isn't hard to write but a bit kludgy-looking.

Reply:
Isn't this just:

```scala
cloned.drop(firstChange).foreach(_ modifyInfo this)
cloned
```

(`modifyInfo` returns `this.type`)

Reply:
Oh! so it is. I've been doing too much FP recently.

- [ ] drop and mutate